basic_pipeline/8_Mixer.md

```diff
@@ -0,0 +1,232 @@
+    Map.update!(state.tracks, pad, fn track ->
+      %Track{track | status: :finished}
+    end)
+
+  state = %{state | tracks: tracks}
+  {{:ok, [{:redemand, :output}]}, state}
+ end
+ ...
+end
+```
+
+What we did here was similar to the logic defined in the `handle_process/4` - we have just updated the state of the track (in that case - by setting its status as `:finished`) and then we called the `handle_demand/5` callback with the `:redemand` actions. The `handle_demand/5` will take care of the fact that the track state has changed.
+There is nothing left to do apart from defining the `handle_demand/5` itself!
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ @impl true
+ def handle_demand(:output, size, _unit, ctx, state) when size > 0 do
+  {state, actions} = update_state_and_prepare_actions(state, ctx.pads)
+  {{:ok, actions}, state}
+ end
+
+ @impl true
+ def handle_demand(_ref, _size, _unit, _ctx, state), do: {state, []}
+ ...
+end
+```
+
+We have split that definition into two parts, depending on the demand size. If the demand size is greater than 0, we are calling the `update_state_and_prepare_actions/2` function (to which we have delegated the whole logic concerning the tracks processing). Otherwise, we do nothing.
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ defp update_state_and_prepare_actions(state, pads) do
+  {state, buffer_actions} = output_buffers(state)
+  {state, end_of_stream_actions} = send_end_of_stream(state)
+  {state, demand_actions} = demand_on_empty_tracks(state, pads)
+
+  actions = buffer_actions ++ end_of_stream_actions ++ demand_actions
+  {state, actions}
+ end
+ ...
+end
+```
+
+The private function `update_state_and_prepare_actions/2` is defined as described in the steps' list above - it aggregates the actions taken in order to:
++ output the ready buffers
++ send `:end_of_stream` notification if necessary
++ demand on empty tracks
+
+Each of these steps has a corresponding private function.
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ defp output_buffers(state) do
+  {buffers, tracks} = prepare_buffers(state.tracks)
+  state = %{state | tracks: tracks}
+  buffer_actions = Enum.map(buffers, fn buffer -> {:buffer, {:output, buffer}} end)
+  {state, buffer_actions}
+ end
+
+ defp prepare_buffers(tracks) do
+  active_tracks =
+    tracks
+    |> Enum.reject(fn {_track_id, track} ->
+      track.status == :finished and track.buffer == nil
+    end)
+    |> Map.new()
+
+  if active_tracks != %{} and
+    Enum.all?(active_tracks, fn {_, track} -> track.buffer != nil end) do
+      {track_id, track} =
+        active_tracks
+    |> Enum.min_by(fn {_track_id, track} -> track.buffer.pts end)
+
+  buffer = track.buffer
+  tracks = Map.put(tracks, track_id, %Track{track | buffer: nil})
+  {buffers, tracks} = prepare_buffers(tracks)
+  {[buffer | buffers], tracks}
+  else
+    {[], tracks}
+  end
+ end
```

Comment:
wait... in this approach with a single buffer on track I believe we shouldn't make a recurrence call to `prepare_buffers` because we will always prepare at most one buffer because of this condition:
 `Enum.all?(active_tracks, fn {_, track} -> track.buffer != nil end)`, right?

Reply:
Well, imagine a situation, that there are two tracks A and B. Track B is empty and a packet with timestamp=2 comes on track A, and later on there comes `:end_of_stream` on track A. Track A is now in `:finished` state, but has a no-nil buffer. `prepare_buffers` is called, but no buffer is produced since track B is empty. 
Then comes a packet with timestamp=1  on track B and later on there comes `:end_of_stream`. The `prepare_buffers` is called once again, and due to `Enum.min_by`, the buffer from track B is chosen. If the `prepare_buffer` wasn't called recursively, nothing would call `prepare_buffers` once again, to transmit the buffer from track A.